test/unit/features/directives/ref.spec.js

```diff
@@ -0,0 +1,93 @@
+import Vue from 'vue'
+
+describe('Directive v-ref', () => {
+  const components = {
+    test: {
+      id: 'test'
+    },
+    test2: {
+      id: 'test2'
+    }
+  }
+
+  it('should accept hyphenated refs', () => {
+    const vm = new Vue({
+      template: `<div>
+        <test v-ref:test></test>
+        <test2 v-ref:test-hyphen></test2>
+      </div>`,
+      components
+    })
+    vm.$mount()
+    expect(vm.$refs.test).toBeTruthy()
+    expect(vm.$refs.test.$options.id).toBe('test')
+    expect(vm.$refs['test-hyphen']).toBeTruthy()
+    expect(vm.$refs['test-hyphen'].$options.id).toBe('test2')
+  })
+
+  it('should accept camelCase refs', () => {
+    const vm = new Vue({
+      template:
+        `<div>
+          <test v-ref:test></test>
+          <test2 v-ref:testCase></test2>
+        </div>`,
+      components
+    })
+    vm.$mount()
+    expect(vm.$refs.test).toBeTruthy()
+    expect(vm.$refs.test.$options.id).toBe('test')
+    expect(vm.$refs.testCase).toBeTruthy()
+    expect(vm.$refs.testCase.$options.id).toBe('test2')
+  })
+
+  it('should accept HOC component', () => {
+    const vm = new Vue({
+      template: '<test v-ref:test></test>',
+      components
+    })
+    vm.$mount()
+    expect(vm.$refs.test).toBeTruthy()
+    expect(vm.$refs.test.$options.id).toBe('test')
+  })
+
+  it('should accept dynamic component', done => {
+    const vm = new Vue({
+      template: `<div>
+        <component :is="test" v-ref:test></component>
+      </div>`,
+      components,
+      data: { test: 'test' }
+    })
+    vm.$mount()
+    expect(vm.$refs.test.$options.id).toBe('test')
+    vm.test = 'test2'
+    waitForUpdate(() => {
+      expect(vm.$refs.test.$options.id).toBe('test2')
+      vm.test = ''
+    }).then(() => {
+      expect(vm.$refs.test).toBeUndefined()
+      done()
+    })
```

Comment:
Removed keep-alive test due to it currently being depreciated. 

Changed the structure to use `.then(done)` syntax. 

All tests are now receiving this error once rebasing from `next`.

```
Chrome 50.0.2661 (Mac OS X 10.11.4) Directive v-ref should accept dynamic component FAILED
	SyntaxError: missing ) after argument list
	    at compileToFunctions (/Users/blake/work/vue/vue/test/unit/index.js:5251:17 <- webpack:///src/entries/web-compiler.js:68:15)
	    at Vue._webRuntime2.default.$mount (/Users/blake/work/vue/vue/test/unit/index.js:117:70 <- webpack:///src/entries/web-runtime-with-compiler.js:35:61)
	    at Object.<anonymous> (/Users/blake/work/vue/vue/test/unit/index.js:8457:9 <- webpack:///test/unit/features/directives/ref.spec.js:62:7)
```